aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -162,10 +171,342 @@ __global__ void gatherTopK(at::cuda::detail::TensorInfo<T, IndexType> input,
+  }
+  __syncthreads();
+
+  items_per_thread = (blk_idx_in_slice + 1 < blocks_per_slice)
+      ? items_per_thread
+      : at::ceil_div((int64_t)(slice_size - blk_idx_in_slice * items_per_block), (int64_t)BLOCK_THREADS);
+
+  // collect digit counts and store in shared memorey
+  for (int i = 0; i < items_per_thread; ++i) {
+    // Find the start offset for this slice
+    IndexType idx = blk_idx_in_slice * items_per_block + i * BLOCK_THREADS + tidx;
+    if (idx < slice_size) {
+      idx *= withinSliceStride;
+      Bitwise val = TopKTypeConfig<T>::convert(doLdg(&data[idx]));
+      bool has_val = ((val & desiredMask) == (desired & desiredMask));
+      Bitwise digit = at::cuda::Bitfield<Bitwise>::getBitfield(val, current_bit, RADIX_BITS);
+      if (has_val) {
+        atomicAdd(&temp_storage.digit_counters[digit], 1);
+      }
+    }
+  }
+
+  __syncthreads();
+
+  // load digit counter to register, one digit per thread
+  static_assert(RADIX_DIGITS <= BLOCK_THREADS, "this kernel requires RADIX_DIGITS <= BLOCK_THREADS");
+  IndexType digit_count = 0;
+  if (tidx < RADIX_DIGITS) {
+    digit_count = temp_storage.digit_counters[tidx];
+  }
+
+  // if blocks_per_slice == 1, there is no need to do cross-block reduction
+  // in this case counts saved at registers instead of global memory
+  if (blocks_per_slice > 1) {
+
+    if (tidx < RADIX_DIGITS) {
+      counts[block_idx * RADIX_DIGITS + tidx] = digit_count;
+    }
+    __threadfence(); // make sure writes are globally visible
+    __syncthreads(); // make sure all writes are finished before update semaphores
+  }
+
+  // the last block of each slice accumulates counters from multiple blocks and updates desired and ks_to_find
+  __shared__ bool s_is_last_block_done;
+
+  if (tidx == 0) {
+    if (blocks_per_slice == 1) {
+      s_is_last_block_done = true;
+    } else {
+      int blocks_finished_old = atomicAdd(&semaphores[slice_idx], 1);
+      s_is_last_block_done = (blocks_finished_old == blocks_per_slice - 1);
+    }
+  }
+
+  __syncthreads();
+
+  if (!s_is_last_block_done)
+    return;
+
+  // accumulates counters from multiple blocks
+  if (tidx < RADIX_DIGITS && blocks_per_slice > 1) {
+    digit_count = 0;
+    for (int blk = 0; blk < blocks_per_slice; ++blk) {
+      digit_count += counts[(slice_idx * blocks_per_slice + blk) * RADIX_DIGITS + tidx];
+    }
+  }
+
+  // compute the block-wide inclusive prefix sum
+  IndexType& digit_count_cumsum = digit_count;
+  BlockPrefixCallbackOp prefix_op(0);
+  BlockScan(temp_storage.scan_storage).InclusiveSum(digit_count, digit_count_cumsum, prefix_op);
+  __syncthreads();
+  // every thread also need the perfix_sum of it's left value for comparison, so save a copy in shared mem
+  if (tidx < RADIX_DIGITS) {
+    temp_storage.digit_count_cumsum[tidx] = digit_count_cumsum;
+  }
+  __syncthreads();
+
+  if (tidx < RADIX_DIGITS) {
+    IndexType digit_count_cumsum_left = (tidx == 0) ? 0 : temp_storage.digit_count_cumsum[tidx - 1];
```

Comment:
What would happen if `k_to_find` is `0`? Look like the following if-clause will not execute. Is this the desired behavior?

Reply:
https://github.com/pytorch/pytorch/pull/71081/files#diff-4b12f751689c3c828c019c6e01c983db4c2929705f93199cf3a6873234d305fcR21-R23

```
if (k == 0) {
    return;
  }
```

the kernel is not launched when `k == 0`, for largest topk when `k == n`, `k_to_find` is `n - k + 1 = 1`, it will not be 0 neither.

Reply:
k_to_find will never reach 0 unless k is 0.

For example, the first pass
```
k_to_find:   2
digit:       0 1 2 3 4
count:       1 1 1 1 1 
cumsum:      1 2 3 4 5
cumsum_left: 0 1 2 3 4
               ↑
```
updated k_to_find:  2 - 1 = 1

next pass
```
k_to_find:   1
digit:       0 1 2 3 4
count:       0 1 0 0 0 
cumsum:      0 1 1 1 1
cumsum_left: 0 0 1 1 1
               ↑
```
from now on the k_to_find will always be: 1 - 0 = 1